lib/liquid/standardfilters.rb

```diff
@@ -89,13 +89,21 @@ def truncate(input, length = 50, truncate_string = "...")
 
     def truncatewords(input, words = 15, truncate_string = "...")
       return if input.nil?
-      wordlist = input.to_s.split
-      words    = Utils.to_integer(words)
+      words = Utils.to_integer(words)
 
-      l = words - 1
-      l = 0 if l < 0
+      words = 1 if words <= 0
+
+      # Scan for non-space characters followed by one or more space characters
+      # `words` times.  Also ignore leading whitespace
+      str = input[/\A[ ]*(?:[^ ]*[ ]+){#{words}}/]
 
-      wordlist.length > l ? wordlist[0..l].join(" ").concat(truncate_string.to_s) : input
+      if str
+        str.strip!                # Remove trailing space
+        str.gsub!(/[ ]{2,}/, " ") # Shrink multiple spaces to one space
```

Comment:
Again, we should match on `\s` for split compatibility.

```suggestion
        str.gsub!(/\s{2,}/, " ") # Shrink multiple spaces to one space
```